xarray/core/coordinates.py

```diff
@@ -796,6 +789,113 @@ def _ipython_key_completions_(self):
         ]
 
 
+class DataTreeCoordinates(Coordinates):
+    """
+    Dictionary like container for coordinates of a DataTree node (variables + indexes).
+
+    This collection can be passed directly to the :py:class:`~xarray.Dataset`
+    and :py:class:`~xarray.DataArray` constructors via their `coords` argument.
+    This will add both the coordinates variables and their index.
+    """
+
+    # TODO: This only needs to be a separate class from `DatasetCoordinates` because DataTree nodes store their variables differently
+    # internally than how Datasets do, see https://github.com/pydata/xarray/issues/9203.
+
+    _data: DataTree  # type: ignore[assignment]  # complaining that DataTree is not a subclass of DataWithCoords - this can be fixed by refactoring, see #9203
+
+    __slots__ = ("_data",)
+
+    def __init__(self, datatree: DataTree):
+        self._data = datatree
+
+    @property
+    def _names(self) -> set[Hashable]:
+        return set(self._data._coord_variables)
+
+    @property
+    def dims(self) -> Frozen[Hashable, int]:
+        # TODO is there a potential bug here? What happens if a dim is only present on data variables?
+        return self._data.dims
+
+    @property
+    def dtypes(self) -> Frozen[Hashable, np.dtype]:
+        """Mapping from coordinate names to dtypes.
+
+        Cannot be modified directly, but is updated when adding new variables.
+
+        See Also
+        --------
+        Dataset.dtypes
+        """
+        return Frozen({n: v.dtype for n, v in self._data._coord_variables.items()})
+
+    @property
+    def variables(self) -> Mapping[Hashable, Variable]:
+        return Frozen(self._data._coord_variables)
+
+    def __getitem__(self, key: Hashable) -> DataArray:
+        if key not in self._data._coord_variables:
+            raise KeyError(key)
+        item = self._data[key]  # type: ignore[index]  # see https://github.com/pydata/xarray/issues/8836
+
+        # TODO perhaps instead use an internal `DataTree` getter method which always returns DataArrays here?
+        return cast("DataArray", item)
+
+    def to_dataset(self) -> Dataset:
+        """Convert these coordinates into a new Dataset"""
+        return self._data.to_dataset()._copy_listed(self._names)
```

Comment:
could use `.dataset` here:
```suggestion
        return self._data.dataset._copy_listed(self._names)
```